clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2230,4 +2238,37 @@ private void throwIfNoAssignorsConfigured() {
                 ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG + " configuration property");
     }
 
+    private MemoryPool createMemoryPool(Metrics metrics, String metricGrpPrefix, long memoryPoolSize, int maxFetchBytes) {
+        if (memoryPoolSize <= 0L || memoryPoolSize <= maxFetchBytes) {
```

Comment:
I think technically the broker is allowed to exceed the max fetch bytes to ensure that at least one message fits into the response. I'm wondering how we would handle that in this patch?

Reply:
The memory pool we create is not strict so it could be used to allocate more than its specified size. The issue is with `maxSingleAllocationBytes`. It's currently set to `maxFetchBytes` so as you pointed it's not going to work. 
Possible solutions:
- Set `maxSingleAllocationBytes` as a factor (x2 ?) of `maxFetchBytes`
- Set `maxSingleAllocationBytes` to max Int